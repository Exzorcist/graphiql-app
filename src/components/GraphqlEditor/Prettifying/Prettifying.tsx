import { PuzzlePieceIcon } from '@heroicons/react/24/outline';
import { useSelector, useDispatch } from 'react-redux';
import { selectRequestValue, changeRequestValue } from '@/redux/slices/graphqlSlice';
import { IGlobalMessage } from '@/types/Message';
import { setMessage } from '@/redux/slices/globalMessageSlice';

import {
  combineRows,
  getRows,
  trimSpaces,
  arrayStringMutation,
  setLineBreak,
  setFieldLine,
  removeDuplicatedSpaces,
  removeBraketSpace,
  addExtraBraketSpace,
  addExtraArgumentSpace,
  correctComma,
  setPadding,
} from './PrettifyingRules';
import { useLocalizationContext } from '@/providers/LocalizationProvider';

function Prettifying() {
  const { t } = useLocalizationContext();
  const dispatch = useDispatch();
  const query = useSelector(selectRequestValue);

  const hanldePrettifying = () => {
    let newQuery: string | string[] = '';

    if (query.includes(';') || query.includes('-') || query.includes('@')) {
      const wrongSymbolMessage: IGlobalMessage = {
        type: 'error',
        text: t.globalMessage.error.prettifying,
        isShown: true,
      };

      dispatch(setMessage(wrongSymbolMessage));
    } else {
      // console.group();
      // console.log(query);
      // console.groupEnd();

      // console.group();
      newQuery = trimSpaces(query);
      newQuery = getRows(newQuery);

      if (Array.isArray(newQuery)) {
        newQuery = arrayStringMutation(newQuery, trimSpaces);
        newQuery = arrayStringMutation(newQuery, removeDuplicatedSpaces);
        newQuery = arrayStringMutation(newQuery, removeBraketSpace);
        newQuery = arrayStringMutation(newQuery, addExtraBraketSpace);
        newQuery = arrayStringMutation(newQuery, addExtraArgumentSpace);
        newQuery = arrayStringMutation(newQuery, correctComma);
        newQuery = setFieldLine(newQuery);
        newQuery = setPadding(newQuery);
        newQuery = arrayStringMutation(newQuery, setLineBreak);
      }

      newQuery = combineRows(newQuery);

      // console.log(newQuery);
      // console.groupEnd();

      dispatch(changeRequestValue(newQuery));
    }
  };

  return (
    <span
      onClick={() => hanldePrettifying()}
      title="Prettify"
      aria-hidden
      className="cursor-pointer opacity-90 transition-colors duration-300 
                 hover:text-editor-accent-light hover:opacity-100"
    >
      <PuzzlePieceIcon className="w-7 h-7" />
    </span>
  );
}

export default Prettifying;
