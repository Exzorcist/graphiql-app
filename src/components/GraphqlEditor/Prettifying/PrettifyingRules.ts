// step 1: Removing spaces at the beginning and end of a line
export const trimSpaces = (str: string): string => str.trim();

// step 2: Get all rows
export const getRows = (str: string): string[] | string => {
  // Search for strings that end in '{' or '}' characters
  const regexStep1 = /[^{}]*[{}]\s*/g;
  const step1 = str.match(regexStep1);

  if (step1 && step1.length) {
    return step1;
  }

  return str;
};

// step 3: Set rows
export const setLines = (str: string): string => `${str}\n`;

// Step 4: Set tabs
// export const setTabsMap = (rows: string[]): string[] => {
//     let count
// };

export const setTabs = (str: string, multiplier: number): string => {
  return `${'  '.repeat(multiplier)}${str}`;
};

// final step: combine rows
export const combineRows = (rows: string[] | string): string => {
  console.log(rows);
  if (Array.isArray(rows)) {
    return rows.join('');
  }

  return rows;
};
