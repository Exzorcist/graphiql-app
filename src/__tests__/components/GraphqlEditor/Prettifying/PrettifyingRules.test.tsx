import { prettifying } from '@/components/GraphqlEditor/Prettifying/PrettifyingRules';

describe('prettifying', () => {
  it('should format the query string', () => {
    const query = `subscription Sub {
        users {id}
      }
      mutation Mut {
        delete_users(where:{id: {_eq:"df"}
        }) {...someFragmant}}
      fragment someFragmant on users_mutation_response {affected_rows returning {id}}`;
    const result = prettifying(query);

    expect(result).toMatchSnapshot();
  });
});
