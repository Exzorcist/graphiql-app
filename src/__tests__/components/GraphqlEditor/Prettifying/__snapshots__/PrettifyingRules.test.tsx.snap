// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prettifying > should format the query string 1`] = `
"subscription Sub {
  users {
    id
  }
}
mutation Mut {
  delete_users(where: { id: { _eq: "df" } }) {
    ...someFragmant
  }
}
fragment someFragmant on users_mutation_response {
  affected_rows
  returning {
    id
  }
}
"
`;
